src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubAppCredentials.java

```diff
@@ -145,20 +185,33 @@ static String generateAppInstallationToken(String appId, String appPrivateKey, S
     @NonNull
     @Override
     public Secret getPassword() {
-        if (Util.fixEmpty(apiUri) == null) {
-            apiUri = "https://api.github.com";
-        }
-
-        long now = System.currentTimeMillis();
         String appInstallationToken;
-        if (cachedToken != null && now - tokenCacheTime < JwtHelper.VALIDITY_MS /* extra buffer */ / 2) {
-            appInstallationToken = cachedToken;
-        } else {
-            appInstallationToken = generateAppInstallationToken(appID, privateKey.getPlainText(), apiUri, owner);
-            cachedToken = appInstallationToken;
-            tokenCacheTime = now;
+        synchronized (this) {
+            try {
+                if (cachedToken == null || cachedToken.isStale()) {
+                    LOGGER.log(Level.FINE, "Generating App Installation Token for app ID {0}", appID);
+                    cachedToken = generateAppInstallationToken(appID,
+                        privateKey.getPlainText(),
+                        actualApiUri(),
+                        owner);
+                    LOGGER.log(Level.FINER, "Retrieved GitHub App Installation Token for app ID {0}", appID);
+                }
+            } catch (Exception e) {
+                if (cachedToken != null && !cachedToken.isExpired()) {
+                    // Requesting a new token failed. If the cached token is not expired, continue to use it.
+                    // This minimizes failures due to occasional network instability,
+                    // while only slightly increasing the chance that tokens will expire while in use.
+                    LOGGER.log(Level.WARNING,
+                        "Keeping cached GitHub App Installation Token for app ID {0}: token is stale but has not expired", appID);
```

Comment:
nit: If we are going to log this as a warning, I think we should include the exception to help diagnose the problem:
```suggestion
                        "Keeping cached GitHub App Installation Token for app ID " + appID + ": token is stale but has not expired", e);
```